test/install/test_install.py

```diff
@@ -0,0 +1,234 @@
+    ./share/orchestra/component_A.json
+    ./some_file
+    ./usr/
+    ./usr/include/
+    ./usr/lib/
+    """
+    ).strip()
+    file_list = tree(orchestra.orchestra_root)
+    assert expected_file_list == file_list
+
+    # Test metadata
+    expected_metadata = {
+        "component_name": "component_A",
+        "build_name": "default",
+        "source": "build",
+        "self_hash": "4c09bf28000a6009576c73ea01cfe74deabb8512",
+        "recursive_hash": "6360092c6e23d3b7611a71920b65246986ee9f91",
+        "binary_archive_path": "component_A/default/none_6360092c6e23d3b7611a71920b65246986ee9f91.tar.gz",
+        "manually_installed": True,
+    }
+    expected_metadata.update(metadata_overrides)
+    metadata = load_json(orchestra.orchestra_root / "share/orchestra/component_A.json")
+    assert_same_metadata(metadata, expected_metadata)
+
+    # Test index (file list)
+    expected_index = dedent(
+        """
+        some_file
+        share/orchestra/component_A.idx
+        share/orchestra/component_A.json
+    """
+    ).strip()
+    with open(orchestra.orchestra_root / "share/orchestra/component_A.idx") as f:
+        filelist = f.read().strip()
+    assert expected_index == filelist
+
+
+def test_install_from_source_with_no_binary_archives_configured(orchestra: OrchestraShim, capsys):
+    """Checks that --fallback-build (-b) causes installation from source if
+    no binary archives repositories are configured"""
+    orchestra("install", "-b", "component_A")
+
+    # Check the install script was run
+    out, err = capsys.readouterr()
+    assert "Executing install script" in out
+
+    check_component_A_installed_properly(orchestra)
+
+
+def test_install_from_source_if_binary_archive_unavailable(orchestra: OrchestraShim, capsys):
+    """Checks that --fallback-build (-b) causes installation from source if
+    the binary archive for the component is not available"""
+    orchestra.add_binary_archive("origin")
+    orchestra("update")
+
+    orchestra("install", "-b", "component_A")
+
+    # Check the install script was run
+    out, err = capsys.readouterr()
+    assert "Executing install script" in out
+
+    check_component_A_installed_properly(orchestra)
+
+
+def test_forced_install_from_source(orchestra: OrchestraShim, capsys):
+    """Checks that --from-source (-B) forces installation from source,
+    even if a binary archive is available"""
+    orchestra.add_binary_archive("origin")
+    orchestra("update")
+    orchestra("install", "-b", "--create-binary-archives", "component_A")
+    out, err = capsys.readouterr()
+    assert "Executing install script" in out
+
+    orchestra.clean_root()
+    orchestra("clean", "--include-sources", "component_A")
+
+    orchestra("install", "-B", "component_A")
+    out, err = capsys.readouterr()
+    assert "Executing install script" in out
+
+    check_component_A_installed_properly(orchestra)
+
+
+def test_install_from_binary_archives(orchestra: OrchestraShim, capsys):
```

Comment:
What's `capsys`?

Reply:
It's a pytest builtin fixture that you can use to capture and read stdin/stderr. I use it in some tests to validate the output or to ensure something did or did not happen (when there's no easier way to do it, e.g. it's one way to assert that a user script run without making it touch the filesystem).